diff --git a/lib/url.c b/lib/url.c
index 42bf1eb..4f4e3f3 100644
--- a/lib/url.c
+++ b/lib/url.c
@@ -592,9 +592,15 @@ CURLcode Curl_init_userdefined(struct UserDefined *set)
 
   /* This is our preferred CA cert bundle/path since install time */
 #if defined(CURL_CA_BUNDLE)
-  result = setstropt(&set->str[STRING_SSL_CAFILE], CURL_CA_BUNDLE);
-  if(result)
-    return result;
+  /* Load CA_BUNDLE, only if it exists */
+  struct stat st;
+  if (lstat(CURL_CA_BUNDLE, &st) == 0) {
+    if (S_ISREG(st.st_mode)) {
+      result = setstropt(&set->str[STRING_SSL_CAFILE], (char *) CURL_CA_BUNDLE);
+      if(result)
+        return result;
+    }
+  }
 #endif
 #if defined(CURL_CA_PATH)
   result = setstropt(&set->str[STRING_SSL_CAPATH], CURL_CA_PATH);
